src/toil/jobStores/azureJobStore.py

```diff
@@ -777,33 +818,45 @@ def retry_on_error(num_tries=5, retriable_exceptions=(socket.error, socket.gaier
     >>> i
     1
     """
-    go = [None]
 
     @contextmanager
     def attempt(last=False):
         try:
             yield
-        except retriable_exceptions as e:
-            # Any instance of these exceptions is automatically retriable.
-            if last:
+        except Exception as e:
+            if expiration is None and last:
+                raise  # This is the single attempt case.
+
+            # Note that delay is set to the next delay not the one that preceded this attempt.
+            elif time.time() + delay > expiration:
+                logger.info('Retry timeout expired, giving up.')
                 raise
-            else:
+            elif type(e) in retriable_exceptions and not last:
                 logger.info("Got a retriable exception %s, trying again" % e.__class__.__name__)
-        except Exception as e:
-            # For other exceptions, the retriable_check function determines whether to retry
-            if retriable_check(e):
+            elif retriable_check(e):
                 logger.info("Exception %s passed predicate, trying again" % e.__class__.__name__)
             else:
                 raise
         else:
             go.pop()
 
-    while go:
-        if num_tries == 1:
-            yield attempt(last=True)
-        else:
-            yield attempt()
-        # It's safe to do this, even with Python's weird default
-        # arguments behavior, since we are assigning to num_tries
-        # rather than mutating it.
-        num_tries -= 1
+    if timeout > 0:
```

Comment:
I don't see why it doesn''t read something like:

try:
      for delay in delays:
              time.sleep(delay)
             rValue, stop = attempt()
             yield rValue
             if stop:
                 break
   else:
          while 1:
               time.sleep(delay)
             rValue, stop = attempt()
             yield rValue
             if stop:
                 break

Longer for sure (but could be cut with nested function), and without:

(1) The horrible side effect in attempt to pop go.
(2) The (I think), fugly use of the iterator. 

But fair enough - I can still read the code, I just don't love it ;)